libp2p/switch.nim

```diff
@@ -75,33 +82,85 @@ type
       streamHandler*: StreamHandler
       secureManagers*: seq[Secure]
       dialLock: Table[PeerID, AsyncLock]
-      ConnEvents: Table[ConnEventKind, HashSet[ConnEventHandler]]
+      connEvents: Table[ConnEventKind, HashSet[ConnEventHandler]]
+      peerEvents: Table[PeerEvent, HashSet[PeerEventHandler]]
 
 proc addConnEventHandler*(s: Switch,
                           handler: ConnEventHandler, kind: ConnEventKind) =
   ## Add peer event handler - handlers must not raise exceptions!
+  ##
+
   if isNil(handler): return
-  s.ConnEvents.mgetOrPut(kind, initHashSet[ConnEventHandler]()).incl(handler)
+  s.connEvents.mgetOrPut(kind,
+    initHashSet[ConnEventHandler]()).incl(handler)
 
 proc removeConnEventHandler*(s: Switch,
                              handler: ConnEventHandler, kind: ConnEventKind) =
-  s.ConnEvents.withValue(kind, handlers) do:
+  s.connEvents.withValue(kind, handlers) do:
     handlers[].excl(handler)
 
 proc triggerConnEvent(s: Switch, peerId: PeerID, event: ConnEvent) {.async, gcsafe.} =
   try:
-    if event.kind in s.ConnEvents:
-      var ConnEvents: seq[Future[void]]
-      for h in s.ConnEvents[event.kind]:
-        ConnEvents.add(h(peerId, event))
+    if event.kind in s.connEvents:
+      var connEvents: seq[Future[void]]
+      for h in s.connEvents[event.kind]:
+        connEvents.add(h(peerId, event))
 
-      checkFutures(await allFinished(ConnEvents))
+      checkFutures(await allFinished(connEvents))
   except CancelledError as exc:
     raise exc
   except CatchableError as exc: # handlers should not raise!
     warn "Exception in triggerConnEvents",
       msg = exc.msg, peerId, event = $event
 
+proc addPeerEventHandler*(s: Switch,
+                          handler: PeerEventHandler,
+                          kind: PeerEvent) =
+  ## Add peer event handler - handlers must not raise exceptions!
+  ##
+
+  if isNil(handler): return
+  s.peerEvents.mgetOrPut(kind,
+    initHashSet[PeerEventHandler]()).incl(handler)
+
+proc removePeerEventHandler*(s: Switch,
+                             handler: PeerEventHandler,
+                             kind: PeerEvent) =
+  s.peerEvents.withValue(kind, handlers) do:
+    handlers[].excl(handler)
+
+proc triggerPeerEvents(s: Switch,
+                       peerId: PeerID,
+                       event: PeerEvent) {.async, gcsafe.} =
+  logScope:
```

Comment:
fwiw, I  stopped using `logScope` because it doesn't always work - ie it'll add to some log messages but not to all

Reply:
huh, ok... thats a fun caveat. 😭